src/odemis/model/_futures.py

```diff
@@ -486,6 +486,64 @@ def set_running_or_notify_cancel(self):
         """
         running = futures.Future.set_running_or_notify_cancel(self)
         if running:
-            self.set_progress(start=time.time())
+            # Set start to current time and end to estimated endtime (difference between current end and start)
+            start, end = self.get_progress()
+            startt = time.time()
+            endt = startt + end - start
+            self.set_progress(start=startt, end=endt)
 
         return running
+
+
+class ProgressiveBatchFuture(ProgressiveFuture):
+    """
+    Representation of a set of ProgressiveFutures which have already been scheduled for execution. The class
+    takes care of time estimates/updates of the batch task. The result is always None.
+    """
+    def __init__(self, futures):
+        """
+        futures (ProgressiveFuture --> float): dict specifying futures and estimated times
+        """
+        self.futures = futures
+        start = time.time()
+        super().__init__(start=start, end=start + sum(self.futures.values()))
+        self.set_running_or_notify_cancel()
+
+        for f in self.futures:
+            f.add_update_callback(self._on_future_update)
+            f.add_done_callback(self._on_future_done)
+
+    def _on_future_update(self, f, start, end):
+        if f.running():  # only care about future which are running, start/end estimates for others not reliable
+            self.futures[f] = end - start
+            self.set_progress(end=self._estimate_end())
+
+    def _on_future_done(self, f):
+        self.set_progress(end=self._estimate_end())
+
+        # Set exception if future failed and cancel all other futures
+        try:
+            ex = f.exception()
+            if ex:
+                self.cancel()
+                self.set_exception(ex)
+        except CancelledError:
```

Comment:
What case are you handling? When .cancel() was just called, and so the (sub-)future is cancelled?  Is that really how it'll happen? I'd expect that in such case `ex` contains a CancelledError, but it'd would not be raised. Or is it that .cancel() would raise CancelledError in such case? If so, please indicate in a comment.